Sources/AsyncAlgorithms/Channels/ChannelStorage.swift

```diff
@@ -0,0 +1,152 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+final class ChannelStorage<Element, Failure: Error>: @unchecked Sendable {
+  private let stateMachine: ManagedCriticalState<ChannelStateMachine<Element, Failure>>
+  private let ids = ManagedCriticalState<Int>(0)
+
+#if DEBUG
+  var onSendSuspended: (() -> Void)?
+  var onNextSuspended: (() -> Void)?
+#endif
+
+  init() {
+    self.stateMachine = ManagedCriticalState(ChannelStateMachine())
+  }
+
+  func generateId() -> Int {
+    self.ids.withCriticalRegion { ids in
+      defer { ids += 1 }
+      return ids
+    }
+  }
+
+  func send(element: Element) async {
+    let producerId = self.generateId()
+    let wasCancelled = ManagedCriticalState(false)
+
+    await withTaskCancellationHandler {
+      await withUnsafeContinuation { (continuation: UnsafeContinuation<Void, Never>) in
```

Comment:
About that, one of my previous implementation of the buffer algorithm (currently in draft PR) involved checking the need to suspend before actually suspending. From a discussion on the forum (https://forums.swift.org/t/clarification-needed-on-unsafecontinuation-documentation/57803/9) it might be counter productive to do that. It seems the caller is not suspended if the continuation is resumed synchronously. 

Checking if we should suspend and then creating the continuation or not might be an overhead VS creating the continuation anyway.

I don’t have the expertise to decide that, perhaps @phausler could jump in?

Reply:
The problem is not soo much about the caller being suspended or not but the actual work of setting up a cancellation handler and creating an unsafe continuation. While the latter is relatively cheap since it is unsafe. The former actually does have a noticeable overhead. So actually checking if we need to suspend or not should bring in performance gains

Reply:
I've added a step where we make an attempt to `send` or to `next`, if we can (we are already waiting for producers or consumers) then we confirm the operation and we don't suspend if we can't we confirm the suspension.